File: Problems/0283_Move_Zeroes/Move_Zeroes.py
# coding: utf-8

import os
import sys
import time
import copy


class Solution:
    def moveZeroes(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        temp = nums[:]
        dst_i = 0
        for src_i in range(len(nums)):
            if temp[src_i] != 0:
                nums[dst_i] = temp[src_i] 
                dst_i += 1
        
        for leaset_i in range(dst_i, len(nums)):
            nums[leaset_i] = 0


    def moveZeroes_old1(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        temp = [0]*len(nums)
        dst_i = 0
        for src_i in range(len(nums)):
            if nums[src_i] != 0:
                temp[dst_i] = nums[src_i] 
                dst_i += 1
        for i in range(len(temp)):
            nums[i] = temp[i]
        '''
        nums = copy.deepcopy(temp)
        '''


def array_str_to_int(flds):
    if len(flds) <= 0:
        return None
    nums = [0]*len(flds)
    nums[0] = int(flds[0])
    for i in range(1, len(flds)):
        nums[i] = int(flds[i])
    return nums


def main():
    argv = sys.argv
    argc = len(argv)

    if (argc < 2):
        print("Usage: python %s <testdata.txt>" %(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("%s not found..." %argv[1])
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        print("argv[1] = %s" %temp)
        loop_main(temp)
    #    print("Hit Return to continue...")
    #    input()


def loop_main(temp):
    tempStr = temp.replace("[","").replace("]","").rstrip()
    flds = tempStr.split(',')
    nums = array_str_to_int(flds)

    print("nums = %s" %nums)

    time0 = time.time()

    sl = Solution()
    sl.moveZeroes(nums)

    print("result = %s" %nums)

    time1 = time.time()
    print("Execute time ... : %f[s]" %(time1 - time0))
    print()


if __name__ == "__main__":
    main()

File: Problems/0200_0299/0283_Move_Zeroes/Project_CS/Move_Zeroes.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Solution
{
    public void MoveZeroes(int[] nums)
    {
        int[] temp = nums;

        int dst_i = 0;
        for (int src_i = 0; src_i < nums.Length; ++src_i) {
            if (temp[src_i] != 0) {
                nums[dst_i] = temp[src_i];
                dst_i++;
            }
        }

        for (int leaset_i = dst_i; leaset_i < nums.Length; ++leaset_i)
            nums[leaset_i] = 0;
        /*
        int leaset_i = dst_i;
        while (leaset_i < nums.Length)
            nums[leaset_i++] = 0;
        */
    }

    private int[] set_array_int(string[] flds)
    {
        if (flds.Length == 0)
            return null;
        
        int[] nums = new int[flds.Length];
        for (int i = 0; i < flds.Length; ++i)
            nums[i] = int.Parse(flds[i]);

        return nums;
    }

    private string output_array_int(int[] nums)
    {
        if (nums.Length <= 0)
            return "";

        string resultStr = nums[0].ToString();
        for (int i = 1; i < nums.Length; ++i)
            resultStr += "," + nums[i].ToString();

        return resultStr;
    }

    public void Main(string args)
    {
        string[] flds = args.Replace("[","").Replace("]","").Split(',');
        int[] nums = set_array_int(flds);
        Console.WriteLine("nums = " + output_array_int(nums));

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        MoveZeroes(nums);
        Console.WriteLine("nums = " + output_array_int(nums));
        
        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms\n");
    }
}
